triples-generator/src/main/scala/io/renku/triplesgenerator/events/consumers/tsprovisioning/triplesuploading/TransformationResultsUploader.scala

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright 2022 Swiss Data Science Center (SDSC)
+ * A partnership between École Polytechnique Fédérale de Lausanne (EPFL) and
+ * Eidgenössische Technische Hochschule Zürich (ETHZ).
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.renku.triplesgenerator.events.consumers.tsprovisioning.triplesuploading
+
+import TriplesUploadResult.NonRecoverableFailure
+import cats.MonadThrow
+import cats.data.EitherT
+import cats.effect.Async
+import cats.syntax.all._
+import io.renku.graph.config.{GitLabUrlLoader, RenkuUrlLoader}
+import io.renku.graph.model.entities.{Person, Project}
+import io.renku.graph.model.{GitLabApiUrl, RenkuUrl}
+import io.renku.jsonld.JsonLD.JsonLDEntity
+import io.renku.jsonld.{JsonLD, NamedGraph}
+import io.renku.jsonld.syntax._
+import io.renku.triplesgenerator.events.consumers.{ProcessingRecoverableError, TSVersion}
+import io.renku.triplesstore.{ProjectsConnectionConfig, RenkuConnectionConfig, SparqlQuery, SparqlQueryTimeRecorder}
+import org.typelevel.log4cats.Logger
+
+private trait TransformationResultsUploader[F[_], TS <: TSVersion] {
+  def execute(query:  SparqlQuery): EitherT[F, ProcessingRecoverableError, Unit]
+  def upload(project: Project):     EitherT[F, ProcessingRecoverableError, Unit]
+}
+
+private object TransformationResultsUploader {
+
+  trait Locator[F[_]] {
+    def apply[TS <: TSVersion](implicit ev: TS): TransformationResultsUploader[F, TS]
+  }
+
+  private class LocatorImpl[F[_]](defaultGraphResultsUploader: TransformationResultsUploader[F, TSVersion.DefaultGraph],
+                                  namedGraphsResultsUploader:  TransformationResultsUploader[F, TSVersion.NamedGraphs]
+  ) extends Locator[F] {
+
+    override def apply[TS <: TSVersion](implicit ev: TS) = ev match {
```

Comment:
just a thought: maybe it is possible to use two overloaded methods accepting the concrete TS - like `def apply(v: TSVersion.DefaultGraph.type) …`? Only to move the selection into compile time :). Ofc, it only works if the call sites know their concrete `TSVersion` - not sure if that applies. The selection could only happens once and then it can continue with the abstract `TransformationResultsUploader`). Maybe we could even use a "compound uploader" that works for both versions (calling both in sequence or in parallel)?

Reply:
That's true, passing the TSVersion implicitly does not give us much here. I'll try to change it.